instrumentation/elasticsearch/elasticsearch-rest-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/ElasticsearchClientAttributeExtractor.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.elasticsearch.rest;
+
+import static io.opentelemetry.instrumentation.api.internal.AttributesExtractorUtil.internalSet;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.network.internal.NetworkAttributes;
+import io.opentelemetry.instrumentation.api.instrumenter.url.internal.UrlAttributes;
+import io.opentelemetry.instrumentation.api.internal.SemconvStability;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import org.elasticsearch.client.Response;
+
+public class ElasticsearchClientAttributeExtractor
+    implements AttributesExtractor<ElasticsearchRestRequest, Response> {
+  private static final Pattern pathPartNamesPattern = Pattern.compile("\\{([^}]+)}");
+  private final Map<String, EndpointPattern> endpointPatternMap = new ConcurrentHashMap<>();
+
+  private static void setServerAttributes(AttributesBuilder attributes, Response response) {
+    InetAddress hostAddress = response.getHost().getAddress();
```

Comment:
Could you elaborate why you chose not to use `NetClientAttributesExtractor` and `NetClientAttributesGetter`

Reply:
TBH, this was quite a struggle when implementing this on finding the best approach.
I'm also not happy with the solution but it's the best I came up with without changing stuff in the core.

The problem is the following:

Both, `NetClientAttributesExtractor` and `NetClientAttributesGetter` have assumptions in their API when / where certain information is available. For example, `getServerAddress()` is always being called in the `onStart()` method of the extractor with only the `Request` as a parameter. However, in this instrumentation the `ServerAddress` is not resolved yet at the time of `onStart()` for that request. (Because the client can have a list of target nodes and resolves at a lower level the actual server it talks to). So, at this level of instrumentation, that information is only available on the `Response` object. So I would need to set this information in the `onEnd()` of the instrumentation. But unfortunately that's not how those extractors work. That's why I decided to implement this logic in a custom / specific extractor for the Elasticsearch client.

I'm open to better ideas. But, I did spent some time thinking about it already (including additional instrumentation points etc.). But that would complicate things unnecessarily.

Reply:
@mateuszrzeszutek ^^^

Reply:
> However, in this instrumentation the ServerAddress is not resolved yet at the time of onStart() for that request. (Because the client can have a list of target nodes and resolves at a lower level the actual server it talks to).

The `server.address`/`server.port` attributes are supposed to return logical server address/port; if there's no one clear target, perhaps they're just not applicable to the Elasticsearch client?
WDYT about implementing the `server.socket.*` attributes instead? You have an `InetAddress` with an IP address of an immediate peer, it seems to suit these attributes.

Reply:
@mateuszrzeszutek 
I followed the [Elasticsearch semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/database/elasticsearch.md). And we actually want the the *logical server address/port* because this is how we would identify the ES instance the client is talking to.

(The usage of the InetAddress is wrong, I need to fix it in my PR. It rather should be the logical Host address on the `org.apache.http.HttpHost` object.)

The problem is just, that the decision / resolution of what is the server address happens "deeper" in the Java ES client code. However, this information is fully available on the Response object. 

So, the question is how important is it to use the `NetClientAttributesExtractor` and would it justify to make the instrumentation more complicated? Or is it acceptable for a specific use case (where the `NetClientAttributesExtractor` interface does not apply) to have a specific extractor that would set the same attributes?

Reply:
> The problem is just, that the decision / resolution of what is the server address happens "deeper" in the Java ES client code that is not suitable for clean instrumentation. However, this information is fully available on the Response object.
> 
> So, the question is how important is it to use the `NetClientAttributesExtractor` and would it justify to make the instrumentation more complicated? Or is it acceptable for a specific use case (where the `NetClientAttributesExtractor` interface does not apply) to have a specific extractor that would set the same attributes?

I see. I think it's fine to have a separate extractor specific to ES then; if there are technical issues with extracting the attributes `onStart()`.
The `NetClientAttributesGettter` (and `ServerAttributesGetter`, for the stable network spec) both have to extract the logical server address and port on start (with just the request) because these two attributes are considered to be sampling relevant for HTTP spans (and I'm guessing that the same will probably apply to other semconvs, like RPC, when they get stabilized). I'd rather not change that just because of one outlier.